Detect new mempool txs

evanlinjin opened the pull request:
### Description

There is a bug in `bdk_bitcoind_rpc` where some new mempool transactions will not be emitted at all.

This problem exists because the avoid-re-emission logic depends on rounded-to-nearest-second timestamps.

The fix is to just emit all mempool transactions but wrap them in `Arc`s so that emission becomes cheap.

**Background:** I tried using `bdk_bitcoind_rpc` as the chain-source to write an example to showcase the [`IntentTracker`](https://github.com/bitcoindevkit/bdk_wallet/pull/257). However, `bdk_bitcoind_rpc` failed to emit some mempool transactions.

### Notes to the reviewers

The test added in c22c68f fails without these fixes.

Some tests are removed as they are no longer relevant.

### Changelog notice

```md
Fixed:
- Some mempool transactions not being emitted at all. The fix is to replace the avoid-re-emission-logic with one which emits all mempool transactions.
```

### Checklists

#### All Submissions:

* [x] I've signed all my commits
* [x] I followed the [contribution guidelines](https://github.com/bitcoindevkit/bdk/blob/master/CONTRIBUTING.md)
* [x] I ran `cargo +nightly fmt` and `cargo clippy` before committing

#### Bugfixes:

* [x] This pull request breaks the existing API
* [x] I've added tests to reproduce the issue which are now passing
~* [ ] I'm linking the issue being fixed by this PR~

evanlinjin committed:
fix(bitcoind_rpc)!: Simplify emitter

Instead of having an avoid-reemission logic based on timestamps returned
by bitcoind RPC, we wrap all transactions in `Arc` and always emit the
entire mempool. This makes emission cheap while avoiding the flawed
avoid-reemission logic.

evanlinjin committed:
test(bitcoind_rpc): Detect new mempool txs

oleonardolima reviewed (commented):
cACK 
I think the new simplified approach looks better.

I'm still in the process of a deeper review. However, when I'm trying to run the `detect_new_mempool_txs`, on both commits, it takes more than 60 seconds and becomes stale, still unsure if it's a local issue.

evanlinjin commented:
@oleonardolima are you on Linux or OSX?

LagginTimes commented:
> cACK I think the new simplified approach looks better.
> 
> I'm still in the process of a deeper review. However, when I'm trying to run the `detect_new_mempool_txs`, on both commits, it takes more than 60 seconds and becomes stale, still unsure if it's a local issue.

FWIW, the `detect_new_mempool_txs` test runs successfully on my machine (Linux).

LagginTimes reviewed (approved):
Re-ACK 51ee99a39fbf9a6dc0ead964e794ca3bebd43ed0

tnull commented on crates/bitcoind_rpc/src/lib.rs:
```diff
@@ -115,110 +116,89 @@ where
+        loop {
+            rpc_tip_height = client.get_block_count()?;
+            rpc_tip_hash = client.get_block_hash(rpc_tip_height)?;
+            rpc_mempool = client.get_raw_mempool_verbose()?;
```
In our experience, we found polling via  `get_raw_mempool_verbose` constantly might be prohibitively expensive when the mempool is full and you 'download' a lot of data from the RPC each time. We since switched to an iterative approach that calls (non-verbose) `getrawmempool` and then only retrieving transaction data via `getmempoolentry` for the transactions that are missing from our in-memory cache (see https://github.com/lightningdevkit/ldk-node/pull/410 , although there have been further optimiziations after that).

Since you're doing the refactoring here already, I wounder if it would it make sense to do a similar change while you're at it.

evanlinjin replied:
@tnull Thank you for the tip.

Changed to use `getrawmempool` without `verbose` in 7de3b71831947d1380c152286af3eb2ab2cdf019.

oleonardolima commented:
> I'm still in the process of a deeper review. However, when I'm trying to run the `detect_new_mempool_txs`, on both commits, it takes more than 60 seconds and becomes stale, still unsure if it's a local issue.

> @oleonardolima are you on Linux or OSX?

It's OSX, but never mind, I ran it from a clean state with 'test-threads=2', and it worked just fine (probably had a concurrency or other problem with other tests I was running using bitcoind).

evanlinjin committed:
feat(bitcoind_rpc)!: Use `getrawmempool` without verbose

This is the more performant method. The downside is that mempool txids
are not returned to the seen-in-mempool timestamps so the caller either
needs to provide this, or we need to rely on std to get the current
timestamp.

* `Emitter::mempool` method now requires the `std` feature. A non-std
  version of this is added: `Emitter::mempool_at` which takes in a
  `sync_time` parameter.

Additional changes:

* `NO_EXPECTED_MEMPOOL_TXIDS` is renamed to `NO_EXPECTED_MEMPOOL_TXS`.
* Updated documentation.
* Fix imports so that `bdk_bitcoind_rpc` compiles without `std`.

evanlinjin committed:
chore(bitcoind_rpc): Make clippy happy

oleonardolima reviewed (approved):
tACK 73ab1eb06bcfaa0368a07b84a054bdd2f4f84f06

It looks good. I've tested in regtest with the bitcoind_rpc example, and it worked just fine.

LagginTimes committed:
docs(bitcoind_rpc): fixed typo in docs

nymius commented:
cACK 51ee99a39fbf9a6dc0ead964e794ca3bebd43ed0
